File: java-to-go-projects/17-kv-store/Command.java
package kvstore;

/**
 * Represents a parsed client command.
 * Supports: GET, SET, DELETE, KEYS, SAVE, LOAD, QUIT
 */
public class Command {

    public enum Type {
        GET, SET, DELETE, KEYS, SAVE, LOAD, QUIT, UNKNOWN
    }

    private final Type type;
    private final String key;
    private final String value;
    private final String raw;

    public Command(Type type, String key, String value, String raw) {
        this.type = type;
        this.key = key;
        this.value = value;
        this.raw = raw;
    }

    public Type getType() { return type; }
    public String getKey() { return key; }
    public String getValue() { return value; }
    public String getRaw() { return raw; }

    /**
     * Parse a raw command string into a Command object.
     *
     * Format: COMMAND key [value]
     * Examples:
     *   SET name Alice
     *   GET name
     *   DELETE name
     *   KEYS
     *   SAVE
     *   LOAD
     *   QUIT
     */
    public static Command parse(String input) {
        String trimmed = input.trim();
        if (trimmed.isEmpty()) {
            return new Command(Type.UNKNOWN, null, null, input);
        }

        String[] parts = trimmed.split("\\s+", 3);
        String cmd = parts[0].toUpperCase();

        return switch (cmd) {
            case "GET" -> {
                if (parts.length < 2) yield error("GET requires a key", input);
                yield new Command(Type.GET, parts[1], null, input);
            }
            case "SET" -> {
                if (parts.length < 3) yield error("SET requires key and value", input);
                yield new Command(Type.SET, parts[1], parts[2], input);
            }
            case "DELETE", "DEL" -> {
                if (parts.length < 2) yield error("DELETE requires a key", input);
                yield new Command(Type.DELETE, parts[1], null, input);
            }
            case "KEYS" -> new Command(Type.KEYS, null, null, input);
            case "SAVE" -> new Command(Type.SAVE, null, null, input);
            case "LOAD" -> new Command(Type.LOAD, null, null, input);
            case "QUIT", "EXIT" -> new Command(Type.QUIT, null, null, input);
            default -> new Command(Type.UNKNOWN, null, null, input);
        };
    }

    private static Command error(String msg, String raw) {
        return new Command(Type.UNKNOWN, null, msg, raw);
    }
}

File: java-to-go-projects/17-kv-store/rust/src/command.rs

pub enum Command {
    Get { key: String },
    Set { key: String, value: String },
    Delete { key: String },
    Keys,
    Save,
    Load,
    Quit,
    Unknown { message: String },
}

pub fn parse(input: &str) -> Command {
    let s = input.trim();

    //if empty, return unknown
    if s.is_empty() {
        return Command::Unknown {
            message: "Empty command".to_string(),
        };
    }

    let parts: Vec<&str> = s.splitn(3, ' ').collect();

    match parts[0].to_uppercase().as_str() {
        "GET" => {
            if parts.len() < 2 {
                return Command::Unknown {
                    message: "GET command requires a key".to_string(),
                };
            }
            Command::Get {
                key: parts[1].to_string(),
            }
        }
        "SET" => {
            if parts.len() < 3 {
                return Command::Unknown {
                    message: "SET command requires a key and value".to_string(),
                };
            }
            Command::Set {
                key: parts[1].to_string(),
                value: parts[2].to_string(),
            }
        }
        "DELETE"|"DEL" => {
            if parts.len() < 2 {
                return Command::Unknown {
                    message: "DELETE command requires a key".to_string(),
                };
            }
            Command::Delete {
                key: parts[1].to_string(),
            }
        }
        "KEYS" => Command::Keys,
        "SAVE" => Command::Save,
        "LOAD" => Command::Load,
        "QUIT"|"EXIT" => Command::Quit,
        _ => Command::Unknown {
            message: format!("Unknown command: {}", s),
        },
    }

}